external/java/src/main/java/ch/epfl/dedis/lib/crypto/bn256/GFpPool.java

```diff
@@ -0,0 +1,72 @@
+package ch.epfl.dedis.lib.crypto.bn256;
+
+import java.util.LinkedList;
+
+/**
+ * This singleton class keeps a pool of different GFp. Instead of creating manually new GFp objects that will then
+ * be eventually deleted by the GC, one can use this class which will keep no longer used objects and re-use them.
+ * If the pool is empty a new object is created. We initialize GFp instances to ZERO.
+ */
+public class GFpPool {
+    private static GFpPool ourInstance = new GFpPool();
+
+    public static GFpPool getInstance() {
+        return ourInstance;
+    }
+
+    private GFpPool() { }
+
+    public GFp2  get2()  { return (GFp2)  get(gfp2Pool);  }
+    public GFp6  get6()  { return (GFp6)  get(gfp6Pool);  }
+    public GFp12 get12() { return (GFp12) get(gfp12Pool); }
+
+    public void put2 (GFp2... gFp2s)   { put(gfp2Pool,  gFp2s);  }
+    public void put6 (GFp6... gFp6s)   { put(gfp6Pool,  gFp6s);  }
+    public void put12(GFp12... gFp12s) { put(gfp12Pool, gFp12s); }
+
+    public int getC2()  { return gfp2Pool.count;  }
+    public int getC6()  { return gfp6Pool.count;  }
+    public int getC12() { return gfp12Pool.count; }
+
+    private GFpItf get(Pool pool) {
+        pool.count++;
+        if(pool.cache.isEmpty())
+            return pool.create();
+
+        return pool.cache.pop().setZero();
```

Comment:
I have a feeling that `setZero` is not needed, could you try to remove it and see if the tests still pass?

If the `setZero` is not essential they maybe you don't need the `GFpItf` interface. You can just put `Object` in the pool (`private void put(Pool pool, Object... gfps)`) and then cast them back to the right type. You can also keep an empty private interface.

Reply:
In my opinion removing the interface and using `private void put(Pool pool, Object... gfps)` is probably better, because the the public methods have the correct typed anyway.